### PR TITLE
HPCC-20643 Move spark build to a plugin

### PR DESCRIPTION
**Pull request by Michael-Gardner**
Signed-off-by: Michael Gardner <michael.gardner@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by Michael-Gardner** [comment]
@xwang2713 Changes to the spark build so that it acts like a plugin.  SPARK=ON will do a spark plugin build.  INCLUDE_SPARK=ON will include spark into a platform build for development purposes, and -DINCLUDE_PLUGINS=ON && SUPPRESS_SPARK=ON will remove spark from a with-plugins build.

**Comment by xwang2713** [comment]
Code looks good. Just one thing with or without plugins the platform package names are the same.
For example, Platform + Spark and Platform + Plugins (include Spark) have the same package name as: hpccsystems-platform-community-with-spark_<version><distro>_<arch>.<deb|rpm>.
Test builds run successfully:
1) Plugin: -DSPARK=ON : http://10.240.32.243/view/custom/job/CE-Candidate-Plugins-Spark-JIRA-20643/
2) platform with Spark  -DINCLUDE_SPARK=ON: http://10.240.32.243/view/custom/job/CE-Platform-with-Spark-20643/
3) platform with all plugins (include Spark) : -DINCLUDE_PLUGINS=ON: http://10.240.32.243/view/custom/job/CE-Platform-with-Plugins-include-Spark-20643/
4) platform with all plugins exclude Spark: DINCLUDE_PLUGINS=ON -DSUPPRESS_SPARK=ON: http://10.240.32.243/view/custom/job/CE-Platform-with-Plugins-no-Spark-20643/

**Comment by richardkchapman** [comment]
@rpastrana Can you approve this?

**Comment by rpastrana** [comment]
@richardkchapman code looks fine, but awaiting test results from Godji

**Comment by richardkchapman** [comment]
@Michael-Gardner Looks like perhaps you forgot to check in some files?

**Comment by Michael-Gardner** [comment]
Yeah I moved stuff around and didn't take an install() entry out somewhere.  Fixing it now.

**Comment by AttilaVamos** [comment]
@Michael-Gardner This startup problem belongs to Smoketest. I will fix it and reschedule this PR.

**Comment by richardkchapman** [comment]
@rpastrana Please review

File: initfiles/componentfiles/configxml/cgencomplist_linux.xml.in [hunk 1]
```diff
@@ -51,7 +51,10 @@
   <Component name="backupnode" processName="BackupNode" schema="backupnode.xsd">
     <File name="backupnode_vars.xsl" method="xslt" destName="backupnode.conf"/>
   </Component>
-  @SPARKTHOR_CGENCOMPLIST_LINUX_ENTRY@
+  <Component name="sparkthor" processName="SparkThor" schema="sparkThor.xsd">
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/cgencomplist_linux.xml.in]: Does this mean sparkthor will always be configurable via configmgr even if the plugin isn't installed?

File: plugins/spark/spark-environment.xml.in [hunk 2]
```diff
@@ -0,0 +1,1250 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Generated with envgen -->
```
Comment (rpastrana) [inline plugins/spark/spark-environment.xml.in]: very minor, but this comment prob doesn't apply anymore, perhaps an updated comment would hep. <-- default environment for such a such build--> ?

File: plugins/spark/spark-environment.xml.in [hunk 3]
```diff
@@ -0,0 +1,1250 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (rpastrana) [inline plugins/spark/spark-environment.xml.in]: is this an exact copy of another environment.xml + sparkthor entries? if so, we shouldn't duplicate the environment.

File: plugins/spark/spark-environment.xml.in [hunk 4]
```diff
@@ -0,0 +1,1250 @@
...
+                         resource="WsEclAccess"
+                         service="ws_ecl"/>
+    <Authenticate access="Read"
+                  description="Root access to WS ECL service"
+                  path="/"
+                  required="Read"
+                  resource="WsEclAccess"/>
+   </Properties>
+  </EspService>
+  @WSSQL_SERVICE_BLOCK_S@
```
Comment (rpastrana) [inline plugins/spark/spark-environment.xml.in]: is this intentional?

**Review by rpastrana — COMMENTED** [review]
@Michael-Gardner looks reasonable. left a few comments.

File: plugins/spark/spark-environment.xml.in (hunk 3 above)
Comment (Michael-Gardner) [reply plugins/spark/spark-environment.xml.in]: Yes, it's just the default xml + the sparkthor stuff.  Should I remove this file then?

**Review by xwang2713 — COMMENTED** [review]
Looks good

**Comment by richardkchapman** [comment]
@Michael-Gardner Please squash

**Comment by Michael-Gardner** [comment]
@richardkchapman Squashed

**Pull request merged**